instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.tomcat7;
+
+import static io.opentelemetry.javaagent.instrumentation.tomcat7.TomcatTracer.tracer;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.servlet.AppServerBridge;
+import io.opentelemetry.javaagent.instrumentation.servlet.v3_0.TagSettingAsyncListener;
+import java.util.concurrent.atomic.AtomicBoolean;
+import net.bytebuddy.asm.Advice;
+import org.apache.coyote.Adapter;
+import org.apache.coyote.Request;
+import org.apache.coyote.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("unused")
+public class ServerHandlerAdvice {
+  private static final Logger log = LoggerFactory.getLogger(ServerHandlerAdvice.class);
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static void onEnter(
+      @Advice.This Adapter itsMe,
+      @Advice.Argument(0) Request request,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    Context attachedContext = tracer().getServerContext(request);
+    if (attachedContext != null) {
+      log.debug("Unexpected context found before server handler even started: {}", attachedContext);
+      return;
+    }
+
+    context = tracer().startServerSpan(request, itsMe.getClass().getSimpleName());
+
+    scope = context.makeCurrent();
+  }
+
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void stopSpan(
+      @Advice.Argument(0) Request request,
+      @Advice.Argument(1) Response response,
+      @Advice.Thrown Throwable throwable,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    if (scope == null) {
+      return;
+    }
+    scope.close();
+
+    if (context == null) {
+      return;
+    }
+
+    Throwable throwableToReport =
+        throwable != null ? throwable : AppServerBridge.getUnhandledThrowable(context);
+
+    if (throwableToReport != null) {
+      if (response.isCommitted()) {
+        tracer().endExceptionally(context, throwableToReport, response);
+      } else {
+        tracer().endExceptionally(context, throwableToReport);
```

Comment:
can you add a comment here about why not passing response when it's not committed? it's not obvious to me at least

Reply:
If the response is not yet committed, then response headers, including response code, are not yet written to the output stream. And the only reason response is passed to the `endExceptionally` is to read the response code.

.... And after looking at the https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/894 and into `endExceptionally` I see that in this case `http.status_code` will be set to 500 even if it wasn't actually sent. Maybe this has to be fixed also, but I'm reluctant to change that logic in the current PR.

Reply:
This comment better be in the code :)

Reply:
Coming soon...